modules/Microsoft.DBforMySQL/flexibleServers/deploy.bicep

```diff
@@ -0,0 +1,406 @@
+param diagnosticEventHubAuthorizationRuleId string = ''
+
+@description('Optional. Name of the diagnostic event hub within the namespace to which logs are streamed. Without this, an event hub is created for each log category.')
+param diagnosticEventHubName string = ''
+
+@description('Optional. The name of logs that will be streamed. "allLogs" includes all possible logs for the resource.')
+@allowed([
+  'allLogs'
+  'MySqlAuditLogs'
+  'MySqlSlowLogs'
+])
+param diagnosticLogCategoriesToEnable array = [
+  'allLogs'
+]
+
+@description('Optional. The name of metrics that will be streamed.')
+@allowed([
+  'AllMetrics'
+])
+param diagnosticMetricsToEnable array = [
+  'AllMetrics'
+]
+
+@description('Optional. The name of the diagnostic setting, if deployed.')
+param diagnosticSettingsName string = '${name}-diagnosticSettings'
+
+var diagnosticsLogsSpecified = [for category in filter(diagnosticLogCategoriesToEnable, item => item != 'allLogs'): {
+  category: category
+  enabled: true
+  retentionPolicy: {
+    enabled: true
+    days: diagnosticLogsRetentionInDays
+  }
+}]
+
+var diagnosticsLogs = contains(diagnosticLogCategoriesToEnable, 'allLogs') ? [
+  {
+    categoryGroup: 'allLogs'
+    enabled: true
+    retentionPolicy: {
+      enabled: true
+      days: diagnosticLogsRetentionInDays
+    }
+  }
+] : diagnosticsLogsSpecified
+
+var diagnosticsMetrics = [for metric in diagnosticMetricsToEnable: {
+  category: metric
+  timeGrain: null
+  enabled: true
+  retentionPolicy: {
+    enabled: true
+    days: diagnosticLogsRetentionInDays
+  }
+}]
+
+@description('Optional. Enable telemetry via a Globally Unique Identifier (GUID).')
+param enableDefaultTelemetry bool = false
+
+var enableReferencedModulesTelemetry = false
+
+resource defaultTelemetry 'Microsoft.Resources/deployments@2021-04-01' = if (enableDefaultTelemetry) {
+  name: 'pid-47ed15a6-730a-4827-bcb4-0fd963ffbd82-${uniqueString(deployment().name, location)}'
+  properties: {
+    mode: 'Incremental'
+    template: {
+      '$schema': 'https://schema.management.azure.com/schemas/2019-04-01/deploymentTemplate.json#'
+      contentVersion: '1.0.0.0'
+      resources: []
+    }
+  }
+}
+
+resource cMKKeyVaultKey 'Microsoft.KeyVault/vaults/keys@2022-07-01' existing = if (!empty(cMKKeyVaultResourceId) && !empty(cMKKeyName)) {
+  name: '${last(split(cMKKeyVaultResourceId, '/'))}/${cMKKeyName}'
+  scope: resourceGroup(split(cMKKeyVaultResourceId, '/')[2], split(cMKKeyVaultResourceId, '/')[4])
+}
+
+resource geoBackupCMKKeyVaultKey 'Microsoft.KeyVault/vaults/keys@2022-07-01' existing = if (!empty(geoBackupCMKKeyVaultResourceId) && !empty(geoBackupCMKKeyName)) {
+  name: '${last(split(geoBackupCMKKeyVaultResourceId, '/'))}/${geoBackupCMKKeyName}'
+  scope: resourceGroup(split(geoBackupCMKKeyVaultResourceId, '/')[2], split(geoBackupCMKKeyVaultResourceId, '/')[4])
+}
```

Comment:
```suggestion
resource cMKKeyVault 'Microsoft.KeyVault/vaults@2022-07-01' existing = if (!empty(cMKKeyVaultResourceId)) {
  name: last(split(cMKKeyVaultResourceId, '/'))
  scope: resourceGroup(split(cMKKeyVaultResourceId, '/')[2], split(cMKKeyVaultResourceId, '/')[4])

  resource cMKKey 'keys@2022-07-01' existing = if (!empty(cMKKeyVaultResourceId) && !empty(cMKKeyName)) {
    name: cMKKeyName
  }

  resource geoBackupCMKKey 'keys@2022-07-01' existing = if (!empty(geoBackupCMKKeyVaultResourceId) && !empty(geoBackupCMKKeyName)) {
    name: geoBackupCMKKeyName
  }
}
```

This should work. Mind you though, we have to update references from for example `geoBackupCMKKeyVaultKey` to `cMKKeyVault::cMKKey`.

We may even be able to remove the 2nd and 3rd `!empty(geoBackupCMKKeyVaultResourceId)` check in the suggestion. But that must be tested as I don't know if Bicep evaluates inner resources if the condition of the outer does not apply.

Reply:
@AlexanderSehr I've applied the suggestion. Will test things out. I 'unresolved' the issue so that I can update here

Reply:
@AlexanderSehr, thinking about this a bit more:

![image](https://user-images.githubusercontent.com/28486158/219806910-8116f463-8791-4a8a-9eea-695eacefd38b.png)

What is the value of having the condition in the inner resource? Even in the previous implementation, the `cMKKeyName` should become a conditional parameter instead of optional, as we need the name of the Key if we are using CMK. But the Key version is the one that is optional. Doesn't this also to other modules implementations we have. Example: 

```bicep
if (!empty(cMKKeyVaultResourceId) && !empty(cMKKeyName))`
```

should be

```bicep
if (!empty(cMKKeyVaultResourceId)`
```

and the deployment should fail if the `cMKKeyName` is not provided. Thoughts?

Reply:
The only reason for the inner condition is that I don't know what Bicep/ARM does if you don't provide it. For example: If no KeyVaultResourceID was provided, does that mean the inner condition isn't evaluated automatically? That would be great. But I wouldn't be surprised if it is anyways and the deployment would fail.

Regarding your point with `Conditional` for the parameter I somewhat agree. However, when viewing the `existing` resources in isolation, I guess the condtions still make sense.

Can you draft a proposal? That would make it easier to discuss :D

Reply:
As discussed. We've moved to the new inner method. I added an issue for aligning other modules as well. See #2838